### ISSUE TITLE
[v1.2] How to configure etcd for https endpoint?

### ISSUE THREAD

**Issue by drnic** [zalando/patroni#384]
I'm enthusiastically trying out new support for etcd https endpoints (plus auth support).

But my output is filled with warnings from urllib3:

```
/usr/lib/python3.5/site-packages/urllib3/connectionpool.py:843: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
InsecureRequestWarning)
```

I've tried installing `pip3 install certifi` but no change. 

I also tried specifying:

```yaml
etcd:
  cacert: /etc/ssl/certs/ca-certificates.crt
```

but this creates an error:

```
2017-01-14 16:07:43,702 INFO: Selected new etcd server https://54.174.22.61:15978
2017-01-14 16:07:43,886 ERROR: Failed to get list of machines from https://54.174.22.61:15978/v2: SSLError(SSLError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:645)'),)
2017-01-14 16:07:43,886 ERROR: Machines cache is empty, no machines to try.
2017-01-14 16:07:43,886 INFO: waiting on etcd
```

To reproduce the error, try using https://www.compose.com/etcd which provides 30-days free trial of an https etcd cluster with root/password. The configuration I'm using looks like:

```yaml
etcd:
  host: portal-nnn.stark-wayne.composedb.com:15000
  protocol: https
  username: root
  password: password
```

**Comment by CyberDem0n** [comment]
Hi,

I think I know what is wrong...
1) Patroni gets a list of etcd client urls either from etcd server or from config file.
2) It resolves DNS names and connects to a resolved IP
3) Certificate is issued for a specific DNS and off course verification is failing, because https://localhost:2379 and https://127.0.0.1:2379 are not the same urls.

It make sense to resolve http://localhost:2379 into http://127.0.0.1:2379, but for https urls must stay unchanged.
I will prepare a fix on Monday.

**Comment by drnic** [comment]
@CyberDem0n thanks xoxo

**Comment by CyberDem0n** [comment]
@drnic Fix is merged into master. We will release a new minor version soon.
But there is one important thing regarding your configuration:

```yaml
etcd:
  cacert: /etc/ssl/certs/ca-certificates.crt
```
Are you sure that `/etc/ssl/certs/ca-certificates.crt` contains the right CA certificate?
Certificate on the https://54.174.22.61:15978/ is issued for aws-us-east-1-portal.15.dblayer.com
```
The certificate is not trusted because the issuer certificate is unknown.
The server might not be sending the appropriate intermediate certificates. 
An additional root certificate may need to be imported.

Error code: SEC_ERROR_UNKNOWN_ISSUER
```

You should definitely either have public CA certificate to check etcd certificates or you should know where to get it.

Once you have it you can either add it to the /etc/ssl/certs/ca-certificates.crt or specify in the `etcd.cacert`

**Comment by drnic** [comment]
Thanks, will try HEAD soon.
That ca reference was just an attempt to get it working; I'll admit that figuring out SSL issues isn't my strong suit. I don't think I have that line in my config currently (outside of this issue).
Thanks for looking into the bug and fixing it! 
Nic






On Wed, Jan 18, 2017 at 5:05 AM -0800, "Alexander Kukushkin" <notifications@github.com> wrote:












@drnic Fix is merged into master. We will release a new minor version soon.

But there is one important thing regarding your configuration:
etcd:
  cacert: /etc/ssl/certs/ca-certificates.crt


Are you sure that /etc/ssl/certs/ca-certificates.crt contains the right CA certificate?

Certificate on the https://54.174.22.61:15978/ is issued for aws-us-east-1-portal.15.dblayer.com
The certificate is not trusted because the issuer certificate is unknown.
The server might not be sending the appropriate intermediate certificates. 
An additional root certificate may need to be imported.

Error code: SEC_ERROR_UNKNOWN_ISSUER



You should definitely either have public CA certificate to check etcd certificates or you should know where to get it.


Once you have it you can either add it to the /etc/ssl/certs/ca-certificates.crt or specify in the etcd.cacert

**Comment by drnic** [comment]
I've upgraded to v1.2.3 and am seeing:

```
/usr/lib/python3.5/site-packages/urllib3/connectionpool.py:852: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
InsecureRequestWarning)
```

Configuration for etcd is:

```
etcd:
  uri: https://root:xxx@xxx.stark-wayne.composedb.com:15978
  host: xxx.stark-wayne.composedb.com
  port: 15900
  protocol: https
  username: root
  password: xxxx
```

(not sure if/which bits of this are required, so am passing thru as much config as I can)

**Comment by CyberDem0n** [comment]
Hi @drnic,

First of all there is not such thing as `etcd.uri`, but `etcd.url`.
Sure, it's possible to mix them, but url will have a higher priority and values extracted from url will be used to replace other values.

It is strange that curl request is not failing, but urllib3 does... Can it be that there is a special config file for curl available in the system?

And to check what is really wrong, can you please do following:
```python
import urllib3
import ssl
http = urllib3.PoolManager(cert_reqs=ssl.CERT_REQUIRED, ca_certs='/path/to/ca.pem')
http.request('GET', 'https://xxx.stark-wayne.composedb.com:15978')
```

Where `/path/to/ca.pem` if the public certificate of the CA signed etcd certificates.
Either it is already somewhere in the system or you should be able to get it from compose guys.

If you find the right CA public certificate, you should put it into Patroni config:
```yaml
etcd:
  cacert: /path/to/ca.pem
```

**Comment by drnic** [comment]
Thanks re `etcd.url` vs `etcd.uri`. Fixed that up.

And thanks for the idea of looking up explicitly where cert file is. Found it in docker image, added it to `etcd.cacert` and warning gone.

For others who find this ticket, I found the cert with `curl -v`:

```
bash-4.3# curl -v https://root:xxx@xxx.stark-wayne.composedb.com:15978/v2/keys
...
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
```
